src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -15,10 +17,14 @@ class PipProvider(AbstractProvider):
     def __init__(
         self,
         factory,  # type: Factory
+        roots,  # type: Set[str]
```

Comment:
Is there a particular reason why this is changed to an argument? The factory always initialises this to an empty set.

Reply:
Because the factory then added to it in `make_requirement_from_install_req`. When passing it to the provider we can't do that - we calculate the set of roots beforehand in the resolver and pass it into the provider.

Basically in `Resolver.resolve`, we check the root set *before* we create the provider, whereas we do it after creating the factory.

I guess we could create the provider first and have a `Provider.add_project_to_roots` method. Is that what you were thinking? I don't really care either way, I guess.

Reply:
I don’t mind either, was just wondering why. The current solution sounds good.